Fix 32 bit compilation

amesgen opened the pull request:
This is a rebased and fixed version of #296 (it does not compile for me on e.g. GHC 9.2 and 9.6). This should resolve #293 and resolve #309. 

This PR also adds i386 CI jobs (also see #207); feel free to remove if you consider this out-of-scope.

(My actual motivation for this PR is the GHC Wasm/WASI backend which is 32bit.)

dcoutts commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -423,7 +424,7 @@ decodeWord64 =
   Decoder (\k -> return (ConsumeWord (\w# -> k (toWord64 w#))))
 #else
-  Decoder (\k -> return (ConsumeWord64 (\w64# -> k (toWord64 w64#))))
+  Decoder (\k -> return (ConsumeWord64 (\w64# -> k (W64# w64#))))
```
I don't understand why we're replacing our conversion helper function `toWord64` with a concrete implementation. And indeed deleting the helper function. Why not just make the helper function do the right thing?

amesgen replied:
Will change, see https://github.com/well-typed/cborg/pull/322/files#r1320790119

dcoutts commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -335,20 +337,19 @@ toWord64 n = W64# (wordToWord64# n)
 toInt64  n = I64# n
 toWord64 n = W64# n
 #endif
-#else
-toInt64  n = I64# (intToInt64# n)
-toWord64 n = W64# (wordToWord64# n)
```
Why are we deleting these rather than fixing them? It's breaking the pattern used below. And it also means we cannot support multiple compiler versions that need differences in these functions.

Surely we should just use CPP here to change the impl of `toWord64` and `toInt64` so that with older GHC versions they're implemented as now, and with newer versions (where the unboxed integer types changed) to use the new implementation (that's inlined at the use site below) of
```
toInt64  n = I64# n
toWord64 n = W64# n
```

amesgen replied:
I think I just removed these compat shims as they were not strictly necessary, but I will add them back for consistency and future-proofing.

dcoutts commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -315,11 +315,13 @@ getDecodeAction (Decoder k) = k (\x -> return (Done x))
 toWord8  :: Word# -> Word8
 toWord16 :: Word# -> Word16
 toWord32 :: Word# -> Word32
+#if WORD_SIZE_IN_BITS == 64
```
For consistency, can we use the same macro as everywhere else: either `ARCH_64bit` or `ARCH_32bit`.

amesgen replied:
Right, will also change the existing usage of `WORD_SIZE_IN_BITS` :+1:

dcoutts commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -986,7 +987,7 @@ type ByteOffset = Int64
 peekByteOffset :: Decoder s ByteOffset
 peekByteOffset = Decoder (\k -> return (PeekByteOffset (\off# -> k (I64#
-#if MIN_VERSION_base(4,17,0)
+#if MIN_VERSION_base(4,17,0) && !defined(ARCH_32bit)
```
Surely this can't be quite the right logic, because 32bit arches prior to some ghc version did need `intToInt64#` didn't they?

amesgen replied:
`PeekByteOffset` always provides an `Int64#` on 32bit:
https://github.com/well-typed/cborg/blob/c8013b3474d876f4da56c869d57e3f3ac7f42dc6/cborg/src/Codec/CBOR/Decoding.hs#L191-L195 so this seems correct to me.

dcoutts replied:
> PeekByteOffset always provides an Int64# on 32bit

which would mean this logic here is wrong. The logic currently says "use intToInt64 if base >= 4.17 and only on 64bit arches".

But what you're saying above is that it should be "always use intToInt64 on 32bit arches", so the combo would be

```
#if MIN_VERSION_base(4,17,0) || defined(ARCH_32bit)
```
so that we use intToInt64 on 32bit arches (for all base versions), and we also use it on all arches from base 4.17 onwards.

Did we check this patch actually compiles on 32bit arches prior to ghc 9.2?

amesgen replied:
> > PeekByteOffset always provides an Int64# on 32bit
> 
> which would mean this logic here is wrong. The logic currently says "use intToInt64 if base >= 4.17 and only on 64bit arches".
> 
> But what you're saying above is that it should be "always use intToInt64 on 32bit arches", so the combo would be
> 
> ```
> #if MIN_VERSION_base(4,17,0) || defined(ARCH_32bit)
> ```
> 
> so that we use intToInt64 on 32bit arches (for all base versions), and we also use it on all arches from base 4.17 onwards.

I think the existing logic is exactly right (it also compiles on all GHC versions >=8.4), as seen eg by a case distinction.

The purpose of this code is to fill the hole `_f` in `PeekByteOffset $ \off# -> k (I64# (_f off#))` (where `k` has input type `ByteOffset ~ Int64`).

 - On 32bit:
   We have `off# :: Int64#` and `I64# :: Int64# -> Int64` on all GHCs. Hence, we can fill `_f = id`.
 - On 64bit:
   We have `off# :: Int#`.
    - On GHC < 9.4, we have `I64# :: Int# -> Int64`.
      Hence, we can fill `_f = id`.
    - Since GHC >= 9.4, we have `I64# :: Int64# -> Int64` (FTR: [corresponding section in the migration guide](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.4#word64int64-use-word64int64) for 9.4).
      Hence, we can fill `_f = intToInt64# :: Int# -> Int64#`

Note that we can fill `_f = id` *except* when we are on GHC >=9.4 and on 64bit, which is exactly the existing condition.

Would be neat if there were some more principled/locally understandable alternative to the current CPP maze :sweat_smile: 

> Did we check this patch actually compiles on 32bit arches prior to ghc 9.2?

Yes, to double check once again, I pushed to a branch in my fork with the second commit ([Support GHC >=9.2 on 32bit](https://github.com/well-typed/cborg/pull/322/commits/299bb3e6918f7e835c6ecac32efe4fada7514d8f)) removed: https://github.com/amesgen/cborg/actions/runs/6183531233
One can see that i686 compilation with all GHCs from 8.4 to 9.0 succeeds (as expected, compilation on >=9.2 fails as the respective commit was removed).
Also, here is the CI run that everything works on all GHCs with all three commits: https://github.com/amesgen/cborg/actions/runs/6138762687

dcoutts commented on cborg/src/Codec/CBOR/Read.hs:
```diff
@@ -510,8 +510,8 @@ go_fast !bs da@(ConsumeNegWord64Canonical k) =
-    DecodedToken sz i@(I64# i#)
-      | isInt64Canonical sz i  -> k i# >>= go_fast (BS.unsafeDrop sz bs)
+    DecodedToken sz (I64# i#)
+      | isInt64Canonical sz i# -> k i# >>= go_fast (BS.unsafeDrop sz bs)
```
I'm confused by this. Did this simply never work?

amesgen replied:
I would have to do some `git blame` archaeology to find out whether this worked in the past due to different type signatures or was just introduced without any testing; even more extremely, the code [didn't even parse on 32bit](https://github.com/well-typed/cborg/pull/322/files#diff-ff3003dee423ae4566843555c0bf4e38aea3df0d209ac5a8c53c58fb64b3bd16L1568-L1571) (introduced in #273) previously, so this code went without any testing since at least then.

dcoutts commented on cborg/src/Codec/CBOR/Read.hs:
```diff
@@ -1796,7 +1796,7 @@ tryConsumeInteger hdr !bs = case word8ToWord hdr of
               sz = 9
 #if defined(ARCH_32bit)
-          in DecodedToken sz (BigIntToken (isWord64Canonical sz (word64ToWord w)) $! toInteger w)
+          in DecodedToken sz (BigIntToken (isWord64Canonical sz w) $! toInteger w)
```
Also looks like this couldn't ever have been type correct.

amesgen replied:
see https://github.com/well-typed/cborg/pull/322/files#r1320791126

dcoutts reviewed (commented):
This would be easier to follow if we could split it into two patches: the first making it compile with older compilers (fixing the things that look like they were just broken and type incorrect), and then a patch to also make it build with newer compilers (that make the changes to the unboxed primitive numeric types).

amesgen commented:
> This would be easier to follow if we could split it into two patches: the first making it compile with older compilers (fixing the things that look like they were just broken and type incorrect), and then a patch to also make it build with newer compilers (that make the changes to the unboxed primitive numeric types).

Good idea, did that in the latest push (also addressed most comments) :+1:

dcoutts commented on cborg/src/Codec/CBOR/Decoding.hs:
```diff
@@ -315,38 +315,46 @@ getDecodeAction (Decoder k) = k (\x -> return (Done x))
 toWord8  n = W8#  n
 toWord16 n = W16# n
 toWord32 n = W32# n
+#if defined(ARCH_64bit)
+toInt64  n = I64# n
+toWord64 n = W64# n
+#endif
+#endif
+#if defined(ARCH_32bit)
+toInt64  n = I64# n
```
It's unclear why this identity change is needed. The definitions in the 64 and 32bit cases here are the same as each other, and they are the same definitions as what where here before this code change.

I was half expecting it to be an identity refactoring in preparation for the next patch, but it's not. Was this intended to be something else? Otherwise it has no effect but makes the logic more complicated.

amesgen replied:
> they are the same definitions as what where here before this code change.

No, they are different for 32bit and GHC >=9.4: Before this change, this meant that we compiled to `toInt64 n = I64# (intToInt64# n)` (incorrect, see https://github.com/well-typed/cborg/pull/322#discussion_r1325736908), and with this change, to `toInt64 n = I64# n` (correct).

But I just realized that this diff belongs into the second, not the first commit; will change that and also try to simplify the logic a bit.

dcoutts reviewed (commented):
I'm still struggling to follow this, and I'm not totally convinced the first patch does compile on 32bit arches. I'd like to try that.

amesgen commented:
> I'm still struggling to follow this, and I'm not totally convinced the first patch does compile on 32bit arches. I'd like to try that.

Yes, I found it very hard to modify these CPP-overflowing modules without constantly double-checking everything by actually compiling it. See https://github.com/well-typed/cborg/pull/322#discussion_r1325736908 at the bottom for CI runs confirming that everything compiles as epxected on various GHCs on 32bit.

phadej commented:
Superseded by https://github.com/well-typed/cborg/pull/351
